packages/Discretization/src/DTK_Interpolation_decl.hpp

```diff
@@ -0,0 +1,469 @@
+/****************************************************************************
+ * Copyright (c) 2012-2018 by the DataTransferKit authors                   *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the DataTransferKit library. DataTransferKit is     *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#ifndef DTK_INTERPOLATION_DECL_HPP
+#define DTK_INTERPOLATION_DECL_HPP
+
+#include "DTK_ConfigDefs.hpp"
+#include <DTK_DetailsBox.hpp>
+#include <DTK_DetailsPoint.hpp>
+#include <DTK_DistributedSearchTree.hpp>
+#include <DTK_FE.hpp>
+#include <DTK_FETypes.h>
+#include <DTK_InterpolationFunctor.hpp>
+#include <DTK_PointSearch.hpp>
+#include <DTK_PointTypes.h>
+#include <DTK_Topology.hpp>
+
+#include <Intrepid2_FunctionSpaceTools.hpp>
+#include <Tpetra_Distributor.hpp>
+
+#include <string>
+
+namespace DataTransferKit
+{
+/**
+ * This class performs an interpolation for a set of given points in a given
+ * mesh.
+ */
+template <typename DeviceType>
+class Interpolation
+{
+  public:
+    /**
+     * Constructor.
+     * @param comm
+     * @param cell_topologies
+     * @param cells vertices associated to each cell
+     * @param cell_nodes_coordinates coordinates of all the nodes in the mesh
+     * @param points_coordinates coordinates in the physical frame of the points
+     * that we are looking for.
+     * @param cell_dof_ids degrees of freedom indices associated to each cell
+     * @param fe_orders degree of the finite elements associated to each cell
+     * @param point_types types of points used to define the finite elements
+     * @param fe_type type of the finite element (DTK_HGRAD, DTK_HDIV, or
+     * DTK_CURL)
+     */
+    Interpolation( Teuchos::RCP<const Teuchos::Comm<int>> comm,
+                   Kokkos::View<DTK_CellTopology *, DeviceType> cell_topologies,
+                   Kokkos::View<unsigned int *, DeviceType> cells,
+                   Kokkos::View<double **, DeviceType> nodes_coordinates,
+                   Kokkos::View<double **, DeviceType> points_coordinates,
+                   Kokkos::View<LocalOrdinal *, DeviceType> cell_dof_ids,
+                   Kokkos::View<unsigned int *, DeviceType> fe_orders,
+                   Kokkos::View<DTK_PointType *, DeviceType> point_types,
+                   DTK_FEType fe_type );
```

Comment:
So for me the big question here is `point_type`. The inputs of `cell_topologies`, `cells`, `nodes_coordinates`, `points_coordinates`, `cell_dof_ids`, and `fe_orders` make sense here. I do agree that naming conventions could be improved and the current doxygen is out of date with the specification.

Given that this is extremely similar to the SILO specification that we started with during the initial design and the mapping to our current interfaces much of this makes sense. (For example see [the silo manual](https://wci.llnl.gov/content/assets/docs/simulation/computer-codes/silo/LLNL-SM-453191.pdf) page 2-95). 

I do think we want to have a separate interface/support for spectral grids (I think this is where `point_types` comes from) instead of rolling it into a single constructor.

Reply:
@Rombur @dalg24 maybe let's start talking about different use cases here that would have different constructors (i.e. those with and without the need for `point_type`) and maybe that will start to clarify the interface?

Reply:
Yes this was the plan. That's why three months ago, I opened an issue to talk about the interface (see [here](https://github.com/ORNL-CEES/DataTransferKit/issues/360)) So we wouldn't change the interface once I am done writing the code...